test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESCloudBasedRepositoryIntegTestCase.java

```diff
@@ -0,0 +1,155 @@
+    public static void stopHttpServer() {
+        httpServer.stop(0);
+        httpServer = null;
+        randomServerErrors = null;
+    }
+
+    @After
+    public void tearDownHttpServer() {
+        if (handlers != null) {
+            handlers.keySet().forEach(context -> httpServer.removeContext(context));
+        }
+    }
+
+    protected abstract Map<String, HttpHandler> createHttpHandlers();
+
+    protected abstract HttpHandler createErroneousHttpHandler(HttpHandler delegate);
+
+    protected static boolean hasRandomServerErrors() {
+        assert randomServerErrors != null;
+        return randomServerErrors;
+    }
+
+    protected static String httpServerUrl() {
+        InetSocketAddress address = httpServer.getAddress();
+        return "http://" + InetAddresses.toUriString(address.getAddress()) + ":" + address.getPort();
+    }
+
+    /**
+     * HTTP handler that injects random service errors
+     *
+     * Note: it is not a good idea to allow this handler to simulate too many errors as it would
+     * slow down the test suite.
+     */
+    @SuppressForbidden(reason = "this test uses a HttpServer to emulate a cloud-based storage service")
+    protected static class ErroneousHttpHandler implements HttpHandler {
+
+        // first key is a unique identifier for the incoming HTTP request,
+        // value is the number of times the request has been seen
+        private final Map<String, AtomicInteger> requests;
+        private final HttpHandler delegate;
+        private final int maxErrorsPerRequest;
+
+        @SuppressForbidden(reason = "this test uses a HttpServer to emulate a cloud-based storage service")
+        protected ErroneousHttpHandler(final HttpHandler delegate, final int maxErrorsPerRequest) {
+            this.requests = new ConcurrentHashMap<>();
+            this.delegate = delegate;
+            this.maxErrorsPerRequest = maxErrorsPerRequest;
+            assert maxErrorsPerRequest > 1;
+        }
+
+        @Override
+        public void handle(final HttpExchange exchange) throws IOException {
+            final String requestId = requestUniqueId(exchange);
+            assert Strings.hasText(requestId);
+
+            final boolean canFailRequest = canFailRequest(exchange);
+            final int count = requests.computeIfAbsent(requestId, req -> new AtomicInteger(0)).incrementAndGet();
+            if (count >= maxErrorsPerRequest || canFailRequest == false || randomBoolean()) {
+                requests.remove(requestId);
+                delegate.handle(exchange);
+            } else {
+                handleAsError(exchange);
+            }
+        }
+
+        private void handleAsError(final HttpExchange exchange) throws IOException {
+            Streams.readFully(exchange.getRequestBody());
+            exchange.sendResponseHeaders(HttpStatus.SC_INTERNAL_SERVER_ERROR, -1);
+            exchange.close();
+        }
+
+        protected String requestUniqueId(final HttpExchange exchange) {
```

Comment:
Maybe better to make this and this class `abstract` and then move this workaround to the GCS version of things. Makes it easier to understand why we're doing this and that we only have to do this because GCS doesn't have some unique ID?

Reply:
Makes perfect sense :+1: